src/storage/txn/txn_status_cache.rs

```diff
@@ -0,0 +1,978 @@
+                    Mutex::new(cache).into()
+                })
+                .collect(),
+            is_enabled: true,
+        };
+        SCHED_TXN_STATUS_CACHE_SIZE
+            .allocated
+            .set(initial_allocated_capacity_total as i64);
+        res
+    }
+
+    pub fn new(capacity: usize) -> Self {
+        Self::with_slots_and_time_limit(
+            TXN_STATUS_CACHE_SLOTS,
+            CACHE_ITEMS_REQUIRED_KEEP_TIME,
+            capacity,
+        )
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test() -> Self {
+        // 1M capacity should be enough for tests.
+        Self::with_slots_and_time_limit(16, CACHE_ITEMS_REQUIRED_KEEP_TIME, 1 << 20)
+    }
+
+    pub fn with_slots_and_time_limit(
+        slots: usize,
+        required_keep_time: Duration,
+        capacity: usize,
+    ) -> Self {
+        Self::new_impl(slots, required_keep_time, capacity, None)
+    }
+
+    /// Create a `TxnStatusCache` instance for test purpose, with simulating
+    /// system time enabled. This helps when testing functionalities that are
+    /// related to system time.
+    ///
+    /// An `AtomicU64` will be returned. Store timestamps
+    /// in milliseconds in it to control the time.
+    #[cfg(test)]
+    fn with_simulated_system_time(
+        slots: usize,
+        requried_keep_time: Duration,
+        capacity: usize,
+    ) -> (Self, Arc<AtomicU64>) {
+        let system_time = Arc::new(AtomicU64::new(0));
+        let res = Self::new_impl(
+            slots,
+            requried_keep_time,
+            capacity,
+            Some(system_time.clone()),
+        );
+        (res, system_time)
+    }
+
+    fn slot_index(&self, start_ts: TimeStamp) -> usize {
+        fxhash::hash(&start_ts) % self.slots.len()
+    }
+
+    /// Insert a transaction status into the cache. The current system time
+    /// should be passed from outside to avoid getting system time repeatedly
+    /// when multiple items is being inserted.
+    ///
+    /// If the transaction's information is already in the cache, it will
+    /// **NOT** be promoted to the most-recent place of the internal LRU.
+    pub fn insert(&self, start_ts: TimeStamp, commit_ts: TimeStamp, now: SystemTime) {
+        if !self.is_enabled {
+            return;
+        }
+
+        let insert_time = now.duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
+        let mut slot = self.slots[self.slot_index(start_ts)].lock();
+        let previous_size = slot.size();
+        let previous_allocated = slot.internal_allocated_capacity();
+        slot.insert_if_not_exist(
+            start_ts,
+            CacheEntry {
+                commit_ts,
+                insert_time,
+            },
+        );
+        let size = slot.size();
+        let allocated = slot.internal_allocated_capacity();
+        // Update statistics.
+        // CAUTION: Assuming that only one TxnStatusCache instance is in a TiKV process.
+        SCHED_TXN_STATUS_CACHE_SIZE
```

Comment:
How about release the mutex hold by `slot` before updating statistics to shorten the critical section?